### PR TITLE
Optimize transactions save, one multiple rows insert

### PR DESCRIPTION
**Pull request by llybin**
Part of #350 should a bit improve speed on all blocks.
Similar https://github.com/burst-apps-team/burstcoin/pull/356

Feel free edit, my first kotlin code.

### CODE REVIEW

File: src/main/kotlin/brs/db/sql/SqlTransactionDb.kt [hunk 1]
```diff
@@ -141,36 +141,62 @@ internal class SqlTransactionDb(private val dp: DependencyProvider) : Transactio
     override fun saveTransactions(transactions: Collection<Transaction>) {
         if (transactions.isNotEmpty()) {
             dp.db.useDslContext<Unit> { ctx ->
```
Comment (llybin) [inline src/main/kotlin/brs/db/sql/SqlTransactionDb.kt]: <Unit> can remove?

File: src/main/kotlin/brs/db/sql/SqlTransactionDb.kt [hunk 2]
```diff
@@ -141,36 +141,62 @@ internal class SqlTransactionDb(private val dp: DependencyProvider) : Transactio
...
+                        transaction.signature,
+                        transaction.timestamp,
+                        transaction.type.type,
+                        transaction.type.subtype,
+                        transaction.senderId,
+                        getAttachmentBytes(transaction),
+                        transaction.blockTimestamp,
+                        transaction.fullHash,
+                        transaction.version,
+                        transaction.message != null,
```
Comment (llybin) [inline src/main/kotlin/brs/db/sql/SqlTransactionDb.kt]: !transaction.message ?

File: src/main/kotlin/brs/db/sql/SqlTransactionDb.kt [hunk 3]
```diff
@@ -140,37 +140,63 @@ internal class SqlTransactionDb(private val dp: DependencyProvider) : Transactio
...
+                    TRANSACTION.FULL_HASH,
+                    TRANSACTION.VERSION,
+                    TRANSACTION.HAS_MESSAGE,
+                    TRANSACTION.HAS_ENCRYPTED_MESSAGE,
+                    TRANSACTION.HAS_PUBLIC_KEY_ANNOUNCEMENT,
+                    TRANSACTION.HAS_ENCRYPTTOSELF_MESSAGE,
+                    TRANSACTION.EC_BLOCK_HEIGHT,
+                    TRANSACTION.EC_BLOCK_ID
+                )
+                transactions.map { transaction ->
```
Comment (harry1453) [inline src/main/kotlin/brs/db/sql/SqlTransactionDb.kt]: `forEach` would be better as it doesn't create a new collection with the map results

File: src/main/kotlin/brs/db/sql/SqlTransactionDb.kt (hunk 2 above)
Comment (harry1453) [reply src/main/kotlin/brs/db/sql/SqlTransactionDb.kt]: this field is hasMessage so needs a null comparison. ! as an operator only works on bools so would not work here

**Comment by llybin** [comment]
> Looks good! Have you run a before / after benchmark?

yes, not a lot transactions on testnet in one block, but should be better when more transactions in block. IndirectTransactions gave more profit than that on testnet.

**Comment by harry1453** [comment]
Okay, I'm not surprised that this would be faster, I think the previous method originates from a time when default values were used but I'm not sure...

**Comment by llybin** [comment]
The previous method used `batch, `batch` functions in Burst create several queries, on each record - query. Also they use UPSERT more complicated than just insert or update, and I think we should to know what to do insert or update, and don't call upsert.
Now only one insert with multi-values. I will try do same with account, there more complicated logic, trying undestand that. Next, I'd like to try optimize selects, maybe it possible to get all accounts in block by 1 select query instead 19k "empty" queries, need try and test it.

https://www.jooq.org/doc/3.13/manual/sql-execution/batch-execution/

**Comment by harry1453** [comment]
`upsert` is for insert OR update depending on whether something with the specified key / index already exists. This means we don't have to select everything to determine whether we do an insert or update. In MySQL/MariaDB and H2 this is achieved under the hood using the `MERGE` statement, whilst in SQLite this is achieved under the hood using `INSERT ... ON CONFLICT UPDATE`

Please see
https://www.jooq.org/doc/3.2/manual/sql-building/sql-statements/merge-statement/
https://www.jooq.org/doc/3.13/manual/sql-building/sql-statements/insert-statement/insert-on-duplicate-key/

SQLite does not support `MERGE` btw.

**Pull request merged**